### PR TITLE
a working custom judge implementation

### PR DESCRIPTION
**Pull request by iceboy233**

### CODE REVIEW

**Comment by iceboy233** [comment]
The problem was _POSIX_C_SOURCE not defined in judge.c, causing fdopen() to be undefined, and for some reason the calling convention does not match...

File: jd4/case.py [hunk 1]
```diff
@@ -90,70 +98,175 @@ def __init__(self, open_input, open_output, time_ns, memory_bytes, score):
...
+                    read_pipe(judge_stdout_file, MAX_STDERR_SIZE),
+                    read_pipe(judge_stderr_file, MAX_STDERR_SIZE),
+                    wait_cgroup(user_cgroup_sock,
+                                user_execute_task,
+                                self.time_ns,
+                                self.memory_bytes,
+                                PROCESS_LIMIT),
+                    wait_cgroup(judge_cgroup_sock,
+                                judge_execute_task,
+                                DEFAULT_TIME_NS,
```
Comment (twd2) [inline jd4/case.py:185]: DEFAULT_TIME_NS is too short for some cases?

**Comment by twd2** [comment]
Remark: interactive cases are not supported.

File: jd4/case.py (hunk 1 above)
Comment (iceboy233) [reply jd4/case.py:185]: Good catch, I think we need configurable time and memory limit for the custom judge.

Comment (twd2) [reply jd4/case.py:185]: I think this configuration is global.

**Review by twd2 — APPROVED** [review]
#58

**Comment by undefined-moe** [comment]
I wonder what's the data format for spj?

**Pull request merged**